File: algorithms/medium/golang/5-longest-palindromic-substring.go
package golang

func longestPalindrome(s string) string {
	longestPalindrome := string(s[0])
	for i := 1; i < len(s); i++ {
		odd := maxPalindrome(s, i, i)
		even := maxPalindrome(s, i-1, i)
		if len(longestPalindrome) < len(odd) {
			longestPalindrome = odd
		}
		if len(longestPalindrome) < len(even) {
			longestPalindrome = even
		}
	}
	return longestPalindrome
}

func maxPalindrome(s string, left, right int) string {
	for left >= 0 && right <= len(s)-1 && s[left] == s[right] {
		left--
		right++
	}
	return s[left+1 : right]
}

File: algorithms/medium/python/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        longestPalindrome = s[0]
        for i in range(1, len(s)):
            oddPalindrome = self.getLongestPalindrome(s, i, i)
            evenPalindrome = self.getLongestPalindrome(s, i - 1, i)
            longestPalindrome = max(longestPalindrome, oddPalindrome, evenPalindrome, key=len)
        return longestPalindrome

    def getLongestPalindrome(self, s: str, left: int, right: int) -> str:
        while 0 <= left and right <= len(s) - 1:
            if s[left] != s[right]:
                break
            left -= 1
            right += 1
        return s[left + 1:right]